grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java

```diff
@@ -1311,27 +1317,132 @@ public void grpclbFallbackToBackendsFromResolver() {
+    // The fallback timeout timer is scheduled only once
     assertEquals(0, fakeClock.numPendingTasks(FALLBACK_MODE_TASK_FILTER));
   }
 
-  private void fallbackTestVerifyUseOfFallbackBackendLists(
+  @Test
+  public void grpclbFallback_balancerLostAfterTimeoutExpired() {
+    subtestGrpclbFallbackConnectionLost(true, false, true);
+  }
+
+  @Test
+  public void grpclbFallback_subchannelsLostAfterTimeoutExpired() {
+    subtestGrpclbFallbackConnectionLost(false, true, true);
+  }
+
+  @Test
+  public void grpclbFallback_allLostAfterTimeoutExpired() {
+    subtestGrpclbFallbackConnectionLost(true, true, true);
+  }
+
+  @Test
+  public void grpclbFallback_allLostBeforeTimeoutExpired() {
+    subtestGrpclbFallbackConnectionLost(true, true, false);
+  }
+
+  private void subtestGrpclbFallbackConnectionLost(
+      boolean balancerBroken, boolean allSubchannelsBroken, boolean brokenAfterTimeoutExpires) {
+    long loadReportIntervalMillis = 1983;
+    InOrder inOrder = inOrder(helper, mockLbService);
+
+    // Create a resolution list with a mixture of balancer and backend addresses
+    List<EquivalentAddressGroup> resolutionList =
+        createResolvedServerAddresses(false, true, false);
+    Attributes resolutionAttrs = Attributes.newBuilder()
+        .set(GrpclbConstants.ATTR_LB_POLICY, LbPolicy.GRPCLB).build();
+    deliverResolvedAddresses(resolutionList, resolutionAttrs);
+
+    assertSame(LbPolicy.GRPCLB, balancer.getLbPolicy());
+    inOrder.verify(helper).createOobChannel(
+        addrsEq(resolutionList.get(1)), eq(lbAuthority(0)));
+
+    // Attempted to connect to balancer
+    assertEquals(1, fakeOobChannels.size());
+    ManagedChannel oobChannel = fakeOobChannels.poll();
+    inOrder.verify(mockLbService).balanceLoad(lbResponseObserverCaptor.capture());
+    StreamObserver<LoadBalanceResponse> lbResponseObserver = lbResponseObserverCaptor.getValue();
+    assertEquals(1, lbRequestObservers.size());
+    StreamObserver<LoadBalanceRequest> lbRequestObserver = lbRequestObservers.poll();
+
+    verify(lbRequestObserver).onNext(
+        eq(LoadBalanceRequest.newBuilder().setInitialRequest(
+                InitialLoadBalanceRequest.newBuilder().setName(SERVICE_AUTHORITY).build())
+            .build()));
+    lbResponseObserver.onNext(buildInitialResponse(loadReportIntervalMillis));
+    // We don't care if runSerialized() has been run.
+    inOrder.verify(helper, atLeast(0)).runSerialized(any(Runnable.class));
+    inOrder.verifyNoMoreInteractions();
+
+    // Balancer returns a server list
+    List<ServerEntry> serverList = Arrays.asList(
+        new ServerEntry("127.0.0.1", 2000, "token0001"),
+        new ServerEntry("127.0.0.1", 2010, "token0002"));
+    lbResponseObserver.onNext(buildInitialResponse());
+    lbResponseObserver.onNext(buildLbResponse(serverList));
+
+    List<Subchannel> subchannels =
+        fallbackTestVerifyUseOfBalancerBackendLists(inOrder, helper, serverList);
+
+    // Let fallback timer expire
```

Comment:
I don't know this code well enough to understand exactly what this is covering, so let me say what the behavior should be here and let you tell me if that's what the code is doing.

I think the behavior we want here is for the fallback timer to start again at the moment that we lose contact with the balancers and backends.  If we don't reestablish a connection with either a balancer or a backend by the time the timer expires, then we use the fallback addresses until we regain contact with the balancer.

Note that the above is different from the initial fallback timer that we had implemented previously.  For example, consider the following scenario:

1. Channel is created and is moved out of idle state (either by starting an RPC or by requesting that the channel start connecting).
2. We start trying to connect to the balancer.  At the same time, we start the fallback timer.
3. If the fallback timer expires before we connect to the balancer, we use the fallback addresses while we continue to attempt to connect to the balancers.
4. We successfully connect to the balancer and receive a serverlist.  If the fallback timer had not previously expired, we cancel it; if it had previously expired, we stop using the fallback addresses.  We then start using the backends that the balancer told us about.
5. Some time later, we lose contact with the balancer.  However, we are still in contact with the backends from the last serverlist the balancer sent us, so we keep using them while we try to reconnect to the balancer.
6. Now we lose contact with all backends from the last serverlist the balancer sent us.  We stat a new fallback timer.
7. If we regain contact with the balancer and get a new serverlist before the fallback timer expires, then we cancel the timer and use the new serverlist.
8. If the timer fires before we regain contact with the balancer, then we switch to using the fallback addresses until we regain contact with the balancer.

In other words, there are two different scenarios in which the fallback timer will be used.  The one that we had previously implemented is a timer that starts when we start trying to contact the balancer.  The new one that we need in this PR will start at the moment when we lose contact with the last backend if we have also lost contact with the balancer.

Reply:
In the current code the timeout is used only for the initial fallback. Later fallbacks are not subject to the timeout, thus will activate immediately after all connections are lost (as long as the initial timeout is passed).

We should discuss and decide which approach we want to take.

My understanding is that the initial timeout is necessary because the client needs time to connect to balancer and receive the server list. It doesn't seem to be the case for the later fallbacks.  Losing connections to the balancer and all backends is pretty unusual, and I don't see a reason why we still have to wait for 10 seconds instead of using the fallbacks immediately.

Reply:
@markdroth I have updated the PR to add the timeout to subsequent fallbacks. Now it matches your description.

Reply:
Sounds great!  Then feel free to merge this PR as soon as Penn reviews it.  Thanks!